viewing_party/party.py

```diff
@@ -1,23 +1,162 @@
+            genre_watched[genre] = 1
+
+    for v in genre_watched.values():
+        if v > most_watched:
+            most_watched = v
+    for k in genre_watched.keys():
+        if genre_watched[k] == most_watched:
+            most_watched_genre = k 
+
+    return most_watched_genre
 
 # -----------------------------------------
 # ------------- WAVE 3 --------------------
 # -----------------------------------------
 
-        
+def get_unique_watched(user_data):
+    user_movies = user_data['watched']
+    friends_movies = user_data['friends']
+    friend_movies_combined = []
+    unique_movies = []
+    
+    for friend_dict in friends_movies:
+        for friend_movie in friend_dict['watched']:
+            if friend_movie not in friend_movies_combined:
+                friend_movies_combined.append(friend_movie)
+
+    for movie in user_movies:
+        if movie not in friend_movies_combined:
+            unique_movies.append(movie)
+    
+    return unique_movies
+
+def get_friends_unique_watched(user_data):
+    user_movies = user_data['watched']
+    friends_movies = user_data['friends']
+    unique_movies = []
+    
+    for friend_dict in friends_movies:
+        for friend_movie in friend_dict['watched']:
+            if friend_movie not in user_movies and friend_movie not in unique_movies:
+                unique_movies.append(friend_movie)
+    
+    return unique_movies
+
+
 # -----------------------------------------
 # ------------- WAVE 4 --------------------
 # -----------------------------------------
+def get_available_recs(user_data):
+    user_movies = user_data['watched']
+    friends_movies = user_data['friends']
+    subscriptions = user_data['subscriptions']
+
+    recommended_movies = []
+
+    for friend_dict in friends_movies:
+        for friend_movie in friend_dict['watched']:
+            if friend_movie not in user_movies and friend_movie['host'] in subscriptions:
+                recommended_movies.append(friend_movie)
+
+    return recommended_movies
+
+
 
 # -----------------------------------------
 # ------------- WAVE 5 --------------------
 # -----------------------------------------
+def get_new_rec_by_genre(user_data):
+    most_watched_genre = get_most_watched_genre(user_data)
+
+    user_movies = user_data['watched']
+    friends_movies = user_data['friends']
+
+    recommended_movies = []
+
+    for friend_dict in friends_movies:
+        for friend_movie in friend_dict['watched']:
+            if friend_movie not in user_movies and friend_movie['genre'] == most_watched_genre:
+                recommended_movies.append(friend_movie)
+
+    return recommended_movies
+
+def get_rec_from_favorites(user_data):
+    friends_movies = user_data['friends']
+    favorites = user_data['favorites']
+
+    recommended_movies = []
+    friends_movie_list = []
+
+    for friend_dict in friends_movies:
+        for friend_movie in friend_dict['watched']:
+            friends_movie_list.append(friend_movie)

```

Comment:
Remember that the get_unique_watched function will return all the movies your friends haven't watched. You could leverage that to remove these nested for loops completely. And then in the for loop below, you simply have to check and see if the movie in favorites is also in the unique_watched list.